File: MDfiles/Data structures and Algorithms with Java Python and Javascript/JavaScript/testdoublelinkedlist.js
console.log('test double linked lists')

function Node (e){

    this.e = e;
    this.next= null;
    this.previous=null
}
function dList(){

    this.head = new Node('head');
    this.find = find;
    this.insert =insert;
    this.add = add;
    this.remove=remove;
    this.findLast=findLast;
    this.update = update;
    this.display=display;
    this.size=0;
}

function find(item){

     var curr = this.head;
     while (curr.e!=item){
         curr = curr.next;
     }
        return curr;
}

function insert(e,item){

    var newNode = new Node(e);
    var curr = this.find(item);
    newNode.next= curr.next;
    newNode.previous = curr;
    curr.next = newNode;
}

function add (e){

    var newNode = new Node(e);
    var curr;
    if (this.head.next==null){
        this.head.next = newNode;
    }else {
        curr = this.head.next;
        while (curr.next){
            curr = curr.next;
        }
        newNode.next= curr.next;
        newNode.previous = curr;
        curr.next = newNode;
        
            
    }

}

function update(e,item){
    
    var newNode      = new  Node(e);
    var curr         = this.find(item);
    console.log('it will update '+curr.e);
    newNode.next     = curr.next;
    newNode.previous = curr.previous;
    curr.previous.next = newNode;

}

function remove(item){

    var curr = this.find(item);
    if (curr.next!=null){
        curr.previous.next = curr.next;
        curr.next.previous = curr.previous;
        curr.next =null;
        curr.previous=null;
    }
    if (curr == this.findLast()){
        curr.next =null;
        curr.previous.next=null;
    }

}
function findLast(){
    var curr = this.head;
    while (curr.next!=null){
        curr = curr.next;
    }
    return curr;
}

function display(){
    var city = '';
    var curr = this.head;
    while (curr.next){
        city +=curr.next.e+' ';
        curr= curr.next;
    }
        console.log(city)
}

var l = new dList();
l.add('A');
l.add('B');
l.add('C');
l.display();
l.remove('B');
l.display();
// l.insert('A','head')
// l.insert('B','A');
// l.insert('C','B')
// l.insert('D','C')
// l.display();
// l.update('A1','C');
// l.display();
//l.remove('A')
//l.display();
//console.log(l.findLast())

File: MDfiles/Data structures and Algorithms with Java Python and Javascript/Python/testdoublelinkedlist.py
print('test_double_linked_list')

def test_doublelinkedlist():
    print('\n')
    print('double linked list!')

    dl = dList()
    # dl.insertNode('A','head')
    # dl.insertNode('B','A')
    # dl.insertNode('C','B')
    dl.add('A')
    dl.add('B')
    dl.add('C')
    dl.add('D')
    dl.display()
    print(dl.size)
    dl.remove('A')
    print(dl.size)
    dl.display()
    # dl.update('D','B1')
    # dl.display()
    # print(dl.size)


class Node():
    def __init__(self, e):
        self.e = e
        self.next = None
        self.pre  = None

class dList():

      def __init__(self):
        self.head = Node('head')
        self.size= 0

      def find(self,item):
          curr = self.head
          while curr.e !=item:
              curr = curr.next
          return curr
        
      def insertNode(self,e,item):

          node = Node(e)
          curr = self.find(item)
          node.next = curr.next
          node.pre  = curr
          curr.next = node
          self.size+=1

      def add(self,e):

          node = Node(e)
          if self.head.next==None:
              self.head.next = node
              node.pre = self.head
          else:
              curr = self.head.next
              while curr.next:
                  curr = curr.next
              node.next = curr.next
              node.pre = curr
              curr.next = node
            
          self.size+=1

      def remove (self,item):

          curr = self.find(item)
         
          if curr.next == None:
             
              curr.next =None
              curr.pre.next=None
              curr.pre = None             
          else:
              
              curr.pre.next = curr.next
              curr.next.pre = curr.pre
              curr.next = None
              curr.pre = None
          self.size-=1


      def update(self,item,e):

          node = Node(e)
          curr = self.find(item)
          node.next = curr.next
          node.pre  = curr.pre
          curr.pre.next = node


      def display(self):
          s=''
          curr = self.head
          while curr.next:
              s+=curr.next.e+' '
              curr = curr.next
          print(s)





